addons/sale_coupon/models/coupon_program.py

```diff
@@ -147,14 +147,13 @@ def _filter_not_ordered_reward_programs(self, order):
         Returns the programs when the reward is actually in the order lines
         """
         programs = self.env['coupon.program']
+        order_products = order.order_line.product_id
         for program in self:
-            if program.reward_type == 'product' and \
-               not order.order_line.filtered(lambda line: line.product_id == program.reward_product_id):
+            if program.reward_type == 'product' and program.reward_product_id not in order_products:
                 continue
-            elif program.reward_type == 'discount' and program.discount_apply_on == 'specific_products' and \
-               not order.order_line.filtered(lambda line: line.product_id in program.discount_specific_product_ids):
+            elif program.reward_type == 'discount' and program.discount_apply_on == 'specific_products' and not any(dp in order_products for dp in program.discount_specific_product_ids):
                 continue
-            programs |= program
+            programs += program
```

Comment:
Not fully sure for this one, cannot guarantee this doesn't change existing behavior (a recordset might contain duplicates).
Please leave the code as is, unless you can provide clear stats on your usecase and the perf gain observed.

Reply:
In this case it has the same result, with a performance gain diff.
`|=` is about 16x slower than `+=`

```py
from timeit import timeit

def op_or():
    partners = env["res.partner"]
    for _id in range(1000):
        partners |= env["res.partner"].browse(_id)

def op_add():
    partners = env["res.partner"]
    for _id in range(1000):
        partners += env["res.partner"].browse(_id)

timeit(lambda: op_or(), number=1000)
# 160.22748156497255

timeit(lambda: op_add(), number=1000)
# 9.568011570023373
```


This method is used to filter out unmatching programs, here:

- https://github.com/odoo/odoo/blob/bb59d8dd7ec2beeb7b898197cfd6f4e4831397f5/addons/sale_coupon/models/sale_order.py#L280
- https://github.com/odoo/odoo/blob/bb59d8dd7ec2beeb7b898197cfd6f4e4831397f5/addons/sale_coupon/models/sale_order.py#L297
- https://github.com/odoo/odoo/blob/bb59d8dd7ec2beeb7b898197cfd6f4e4831397f5/addons/sale_coupon/models/sale_order.py#L306-L307

Conceptually, it's the same than `filtered` with lots of conditions. We should not expect nor care about duplicates here IMO.